📖 Proposal external CA and cert sign (#12739)
kubernetes-sigs/cluster-api: docs/proposals/20250908-external-certificate-signing.md

- This feature does not replace existing certificate management solutions such as cert-manager; rather, it provides a mechanism to integrate them.
- This proposal does not modify Kubernetes internal certificate signing behavior beyond enabling delegation to an external authority.

## Proposal

To support external certificate signing and fully external CAs in CAPI, this proposal introduces a spec-level flag in the Cluster resource, e.g., externalCA, externalCAKey, or externalCertSigning. This flag will indicate that the cluster should operate in "external CA mode."

Key aspects of the proposal:

### External CA mode

When the flag is enabled, CA keys are no longer required in the CAPI management cluster and must not be stored or validated in secrets.
CAPI will not generate or manage CA private keys. The CA certificates needs to be provided by the user in Kubernetes secrets with the current format:

- <cluster-name>-etcd
- <cluster-name>-ca
- <cluster-name>-proxy
  CAPI should wait for the CA certificates if they are not already provided. CAPI must not adopt these Secrets, as they are the responsibility of the user.”

### Certificates necessary to Cluster API

> Certificates such as the etcd client certificate and the admin certificate for the kubeconfig are normally generated by CAPI. In external CA mode, these certificates are still needed, but their provisioning and lifecycle management become the responsibility of the user. Users can leverage their own certificate management system, such as cert-manager or any other certificate management solution, to sign and rotate these certificates.

Author: The client certificate keys remain stored in Kubernetes Secrets even in external CA mode. This proposal only allows CAPI to operate without storing, using or validating the CA keys in the management cluster, cloud init or in a workload cluster.

> Certificates such as the etcd client certificate and the admin certificate for the kubeconfig are normally generated by CAPI. In external CA mode, these certificates are still needed, but their provisioning and lifecycle management become the responsibility of the user. Users can leverage their own certificate management system, such as cert-manager or any other certificate management solution, to sign and rotate these certificates.

Author: The user must provide the client certificates directly on the nodes using their preferred mechanism.

The etcd client certificate should be provided by the user in a secret named <cluster-name>-apiserver-etcd-client
The kubeconfig admin certificate should be provided by the user in a secret named <cluster-name>-admin

### Kubeadm Bootstrap: ClusterConfiguration and InitConfiguration

When using Kubeadm bootstrap, the command kubeadm certs generate-csr requires a configuration file containing ClusterConfiguration and/or InitConfiguration. Currently these files are only provided to the initialization control plane.

> These configurations must be provided to joining control plane nodes because they contain the necessary information to generate Certificate Signing Requests (CSRs) for their components. The joining control plane nodes will generate CSRs locally and submit them to the external CA or user-provided signing system.

Author: The init/ClusterConfiguration for joining control plane nodes would be transferred using the same mechanism as for the initial control plane node (using cloud init files)

The responsibilities of CAPI end at enabling external CA mode and signaling that CA keys are managed externally. Generation of CSRs from the nodes and their submission to an external signer are fully the responsibility of the user.

Users can leverage their own mechanisms, such as pre-kubeadm scripts, cloud-init, or any other automation, to generate the CSRs, submit them to the external CA, and retrieve the signed certificates. 

This approach ensures that CAPI remains agnostic to the external signing process, while giving users full flexibility to implement their desired certificate management workflow.

> These configurations must be provided to joining control plane nodes because they contain the necessary information to generate Certificate Signing Requests (CSRs) for their components. The joining control plane nodes will generate CSRs locally and submit them to the external CA or user-provided signing system.

Reviewer: i see, so this CAEP proposes to disable the internal use of CA keys in CAPI by enabling `externalCA`, but the CSR generation, signing of certs and keys and writing them in secrets will be completely external. above, i also ask who is going to write them from secrets to node disks.

Below the workflow that describe this proposal.

![workflow](images/external-cert-sign/workflow.png)

### User Stories

#### Enterprise PKI Integration

As a Cluster API user in an enterprise environment, I want to delegate certificate signing to our corporate CA, so that CA keys never reside in the management cluster and certificate issuance complies with organizational security policies.

#### Flexible Bootstrap Provider Support

As a Cluster API user I want that cluster-api provide all necessary configuration to bootstrap providers to be compatible with external signing, so that I can maintain secure certificate flows regardless of the underlying bootstrap mechanism.

### Security Model

The External CA / External Certificate Signing feature is designed to minimize exposure of CA private keys and improve the overall security posture of Cluster API clusters. The security model includes the following key points:

- CA private keys never leave the external CA system.
- Cluster API and management clusters do not store CA keys in secrets.

## Upgrade Strategy

> No special upgrade plan is required for this feature. Enabling external CA and external certificate signing does not break existing Cluster API behavior. Clusters that use the default internal CA workflow will remain fully functional.

Reviewer: here you'd have to cover how a cluster can enabled the feature on upgrade and roll it back on failure.

Author: Since CAPI does not manage CA rotation automatically, the externalCA feature must be treated as immutable. Enabling it on an existing cluster is not supported. In the future, if CAPI adds support for automated CA rotation, it may be possible to make this field mutable, allowing safe transitions between internal and external CA workflows.

> 

Author: I didn't think about switching from internal CA to external CA. If you want to make this change, you'll need to ask yourself whether you want to keep the same CAs or use new ones. If you want to use new ones (I think this is the case for you, as your current CAs may already have been exposed), you will need to perform a rotation, which is not trivial: https://kubernetes.io/docs/tasks/tls/manual-rotation-of-ca-certificates/

---
title: Fully External CA and external certificate signing in Cluster API
authors:
  - "@moh2a"
reviewers:
  - ""
creation-date: 2025-09-07
last-updated: 2025-09-08
status: provisional
---

# Title

Enabling Cluster API to integrate with fully External CA and delegate certificate signing

## Table of Contents

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Glossary](#glossary)
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals/Future Work](#non-goalsfuture-work)
- [Proposal](#proposal)
  - [User Stories](#user-stories)
    - [Story 1](#story-1)
    - [Story 2](#story-2)
  - [Security Model](#security-model)
- [Upgrade Strategy](#upgrade-strategy)
- [Implementation History](#implementation-history)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Glossary

Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).

### External CA

> An External Certificate Authority (CA) is a CA that exists outside the Cluster API (CAPI) management plane. Instead of having Cluster API generate and manage its own self-signed CA for cluster components (such as kube-apiserver, kubelet, and etcd), an external CA can be integrated to provide trusted root certificates. This is typically used in environments where certificate lifecycles are centrally managed for compliance, auditing, or security policy reasons.

Reviewer 2: The ca is not providing trusted root certs it has to provide all certs that we generate from these cas right?

### External certificate signing

> External certificate signing is the process of delegating certificate issuance to a signing service instead of relying on self-signed certificates generated by Cluster API or a [bootstrap provider](https://cluster-api.sigs.k8s.io/reference/glossary.html#bootstrap-provider). With this approach, the Cluster API user becomes responsible for providing the signed certificates, ensuring their signing by the external CA, and managing their rotation over time.

Reviewer 3: I think we should not talk about " the Cluster API user becomes responsible for providing the signed certificates", because in many plaform/K8s distribution built on top of Cluster API automation is paramount (there should not be human tasks, especially when those iterations might require involvement of different teams/people with different responsibilities)

What about instead using the following wording: "Cluster API delegates to external systems (or users) the responsibility for providing the signed certificates"

(same comments applies to other parts in the doc, I think we should align all of them to the idea that a system or a human can take care of tasks not performed by humans)

## Summary

> This proposal introduces the ability for CAPI to integrate with fully external Certificate Authorities (CAs), meaning that CA key never leaves the external system. It also supports external certificate signing, where all certificate issuance is delegated externally. Cluster components generate Certificate Signing Requests (CSRs) that are signed by the external CA, providing a centralized and consistent trust chain across cluster bootstrap and lifecycle operations.

Reviewer 3: Am I correct is say that this proposal aims to enable external certificate signing for external CA?

This will make a direct link to the glossary above, and drop the need to introduce the concept of "fully" external CA. In other words, at the end CAPI will support

- External CA without external certificate signing  (already supported today)
- External CA with external certificate signing  (this proposal)

With this approach, the CAPI user is responsible for setting up a system to provide signed certificates, coordinate signing with the external CA, and manage certificate rotation over time. This enables integration with enterprise PKI systems, enforces organizational security policies, and ensures compliance with regulatory requirements while maintaining secure communication between cluster components.

## Motivation

Currently, CAPI generates and manages CA keys within the management cluster. These keys are also distributed to control plane nodes via cloud-init userdata during bootstrap. This design exposes several security risks:

- Key exposure in the management cluster: Anyone with access to the management cluster's Secrets has access to the CA keys.
- Key transit over cloud-init: The CA private keys travel over cloud-init user data during control plane provisioning, which could be intercepted or leaked.
- Privilege escalation risk: With access to CA keys, an attacker can sign admin or superadmin certificates, gaining full control over clusters and workloads.

> For reference, native Kubernetes clusters generate CA keys and store them on control plane nodes. This also represents a security risk because compromise of a control plane node could allow signing of highly privileged certificates.

Author: By 'native Kubernetes clusters', I refer to standard clusters deployed with kubeadm, where CA keys are generated and stored locally on the control plane nodes by default. I have [updated ](https://github.com/kubernetes-sigs/cluster-api/pull/12739/changes#diff-52fa92389a92fc1c77afaad2be03648839f99ff4e01030afffd789f706986a08R65)the proposal to avoid any confusion.

By integrating fully external CAs and delegating certificate signing externally:

- CA keys never leave the external system, removing the risk of compromise through the management cluster or cloud-init.
- Cluster API users can implement centralized enterprise PKI practices, ensuring strict access control and auditability for certificate issuance.
- Overall trust in cluster bootstrap and lifecycle operations is increased, mitigating a high-impact attack vector that currently exists in both Cluster API and native Kubernetes.

This proposal addresses a critical security gap, providing users with a mechanism to eliminate CA key exposure while maintaining full Kubernetes functionality.

### Goals

- Enable Cluster API to integrate with fully external Certificate Authorities (CAs) where CA private keys never leave the external system.

> - Support external certificate signing for cluster components (control plane and nodes), delegating certificate issuance to an external CA or signing service.

Reviewer 3: I would phrase this "Make external certificate signing for cluster components possible", because external components/workflow are required to fully support this

> - Allow users to implement their own PKI workflow, including certificate provisioning, signing, and rotation, while maintaining secure cluster bootstrap and lifecycle operations.

Author: Kubelet Certificates: For worker nodes, the Kubelet natively handles its own certificate lifecycle by generating a Certificate Signing Request (CSR) when certificates are nearing expiration. This makes an external custom signer the appropriate mechanism for this layer (https://github.com/kubernetes-sigs/cluster-api/pull/12739#discussion_r2792559626).

Control Plane Certificates: For all control plane components, KubeadmControlPlane (KCP) already supports automated rotation through machine rollouts. This ensures that certificates are refreshed during the standard node lifecycle operations, as documented in the [CAPI certificate rotation guide](https://cluster-api.sigs.k8s.io/tasks/certs/auto-rotate-certificates-in-kcp#automatically-rotating-certificates-using-kubeadm-control-plane-provider).

- Reduce the attack surface related to CA key exposure during cluster bootstrap, management, and rotation.
- Maintain compatibility with existing bootstrap providers while using external signing.

### Non-Goals/Future Work

> - Cluster API will not provide a built-in external CA service; users remain responsible for setting up and managing the external CA.

Reviewer 2: TODO: non-goal get rid of kubeconfig secret + etcd client cert/key secret

- This feature does not replace existing certificate management solutions such as cert-manager; rather, it provides a mechanism to integrate them.
- This proposal does not modify Kubernetes internal certificate signing behavior beyond enabling delegation to an external authority.

## Proposal

To support external certificate signing and fully external CAs in CAPI, this proposal introduces a spec-level field `externalCA` in the Cluster resource. This field indicates that the cluster should operate in "external CA mode" and is gated by a feature flag to track its progression from Alpha → Beta → GA.

```
// externalCA enables external CA mode for the cluster.
// When enabled, CA keys are not stored or validated in the management cluster's secrets.

> ExternalCA *bool `json:"externalCA,omitempty"`

Reviewer 2: I think we should not use the feature flag to enable/disable this behavior for all Clusters.
Usually feature gates are eventually locked to true before they are removed.
That would not be an option here if we also use it to enable/disable it for all Clusters

+ If we scope this feature down to KCP/CABPK this should probably move to the KCP object/controller

> ExternalCA *bool `json:"externalCA,omitempty"`

Reviewer 3: Wandering a little bit if the system can infer this by checking if CA key are provided, similar to what kubeadm does when looking at the etc/kubernetes/pki folder

> ExternalCA *bool `json:"externalCA,omitempty"`

Reviewer 3: I'm leaning towards avoiding API changes and inferring external CA mode from the missing key

```

Key aspects of the proposal:

### External CA mode

When the flag is explicitly enabled, CA keys are no longer required in the CAPI management cluster and must not be stored or validated in secrets. By default, this feature remains disabled to avoid impacting users who do not rely on external CAs.
CAPI will not generate or manage CA private keys. The CA certificates needs to be provided by the user in Kubernetes secrets with the following naming convention:

- <cluster-name>-etcd
- <cluster-name>-ca
- <cluster-name>-proxy

Each Secret should follow this Kubernetes format:
```
apiVersion: v1
kind: Secret
metadata:
  name: <cluster-name>-<etcd|ca|proxy>
type: Opaque
data:
  ca.crt: <base64-encoded-CA-certificate>
```

>   CAPI should wait for the CA certificates if they are not already provided. CAPI must not adopt these Secrets, as they are the responsibility of the user.”

Reviewer 4: ```suggestion
  CAPI should wait for the CA certificates if they are not already provided. CAPI must not adopt these Secrets, as they are the responsibility of the user.
```

### Certificates necessary to Cluster API

Certificates such as the etcd client certificate and the admin certificate for the kubeconfig are normally generated by CAPI. In external CA mode, these certificates are still needed, but their provisioning and lifecycle management become the responsibility of the user. Users can leverage their own certificate management system, such as cert-manager or any other certificate management solution, to sign and rotate these certificates.

The etcd client certificate should be provided by the user in a secret named <cluster-name>-apiserver-etcd-client
The kubeconfig admin certificate should be provided by the user in a secret named <cluster-name>-admin

> 

Reviewer 2: What about the -kubeconfig secret that almost all our controllers depend on?

Author: KCP/CABKP can build it using the secret <cluster-name>-admin

Reviewer 3: Can't we simplify this and let the external tool to manage the kubeconfig secret, including also certificate rotation?

> ### Kubeadm Bootstrap: ClusterConfiguration and InitConfiguration

Reviewer 2: +1 if kubeadm join is able to generate certs with only a JoinConfiguration this should be possible (might require kubeadm changes)

> ### Kubeadm Bootstrap: ClusterConfiguration and InitConfiguration

Reviewer 3: The proposal doesn't give any indication on how certification generation on nodes should happen, we need more clarity before taking any decisions.

Please note that sending ClusterConfiguration to the machines is not easy as it seems, because this will require deep and complex investigations on how this might impact how those file should look like and how to handle them when detecting if a rollout is needed or when to perform in-place upgrades (plus the capability to handle file generate before and after this change).
TL;DR; this might impact the size and scope of this proposal in a relevant way

> When using Kubeadm bootstrap, the command kubeadm certs generate-csr requires a configuration file containing ClusterConfiguration and/or InitConfiguration. Currently these files are only provided to the initialization control plane.

Reviewer 2: ToDo: check current state

- Key transit over cloud-init: The CA private keys travel over cloud-init user data during control plane provisioning, which could be intercepted or leaked.
- Privilege escalation risk: With access to CA keys, an attacker can sign admin or superadmin certificates, gaining full control over clusters and workloads.

In a default configuration, Kubernetes with kubeadm generates generate and store CA keys locally on the control plane nodes. This also represents a security risk because compromise of a control plane node could allow signing of highly privileged certificates.

By integrating external CAs and delegating certificate signing externally:

- CA keys never leave the external system, removing the risk of compromise through the management cluster or cloud-init.
- KCP/CABPK users can implement centralized enterprise PKI practices, ensuring strict access control and auditability for certificate issuance.
- Overall trust in cluster bootstrap and lifecycle operations is increased, mitigating a high-impact attack vector that currently exists in default Kubernetes deployments and, by extension, Cluster API workload cluster.

> This proposal addresses a critical security gap, providing users with a mechanism to eliminate CA key exposure while maintaining full Kubernetes functionality.

Reviewer 3: nit
```suggestion
This proposal provides users with a mechanism to eliminate CA key exposure while maintaining full Kubernetes functionality.
```

I will also move this at ~L66 (before enumerating the advantages a user can achieve by eliminating CA exposure)

### Goals

> - Enable KCP/CABPK to integrate with external Certificate Authorities (CAs) where CA private keys never leave the external system.

Reviewer 3: This goal should be stated in a more generic way, because KCP/CABK are pluggable bootstrap/control plane provider and their behaviour and interactions with CAPI should be defined by the CAPI contral

Reviewer 3: Same comment apply in a few other places in this doc.
We should solve the problem in a generic way, and move all the references to KCP, CABPK (and kubeadm) int a single paragraph where we explain how the generic approach apply to those components

- Make possible to use external certificate signing for cluster components (control plane and nodes), delegating certificate issuance to an external CA or signing service. This allows users to implement their own PKI workflow, including certificate provisioning, signing, and rotation, while maintaining secure cluster bootstrap and lifecycle operations.
- Reduce the attack surface related to CA key exposure during cluster bootstrap and management.

### Non-Goals/Future Work

- Kubeadm will not provide a built-in external CA service; users remain responsible for setting up and managing the external CA.
- This feature does not replace existing certificate management solutions such as cert-manager; rather, it provides a mechanism to integrate them.
- This proposal does not modify Kubernetes internal certificate signing behavior beyond enabling delegation to an external authority.
- The automated rotation of the Certificate Authorities themselves is not in scope for this proposal.
- This proposal does not protect against all damages if an attacker gains access to the management cluster, control plane nodes, or cluster admin credentials.

> - This proposal does not protect against all damages if an attacker gains access to the management cluster, control plane nodes, or cluster admin credentials.

Reviewer 3: ```suggestion
```
duplicated

- The management of the Service Account (SA) key pair is not covered by this proposal and remains handled by the existing logic.

## Proposal

To support external CAs in KCP/CABPK, this proposal introduces a spec-level field `externalCA` in the KubeadmControlPlane resource. This field indicates that the cluster should operate in "external CA mode" and is gated by a feature flag to track its progression from Alpha → Beta → GA.

```
// externalCA enables external CA mode for the cluster.
// When enabled, CA keys are not stored or validated in the management cluster's secrets.
ExternalCA *bool `json:"externalCA,omitempty"`
```

Key aspects of the proposal:

### Affected Certificates Overview

The following table summarizes the certificates impacted by the external CA mode:

> | Certificate / Key Pair    | Responsibility  | Scope         | Location           | Logic / Mechanism                                |

Reviewer 3: I would drop scope, because IMO it is confusing

e.g. KCP provides certificates in compliance to the [Cluster API contract](https://cluster-api.sigs.k8s.io/developer/providers/contracts/control-plane#cluster-certificate-management), but it seems that only the KubeConfig is managed according to the contract (which is not correct)

| :------------------------ | :-------------- | :------------ | :----------------- | :----------------------------------------------- |
| **Cluster CA**            | External system | KCP           | Management Cluster | Key in external system; Cert in Secret           |
| **Etcd CA**               | External system | KCP           | Management Cluster | Key in external system; Cert in Secret           |
| **Front Proxy CA**        | External system | KCP           | Management Cluster | Key in external system; Cert in Secret           |
| **Etcd Client Cert**      | External system | KCP           | Management Cluster | Provided via Secret (for stacked etcd)           |
| **Kubeconfig Admin Cert** | External system | CAPI Contract | Management Cluster | Provided via Secret                              |
| **Controlplane Certs**    | External system | CABPK         | Controlplane node  | Provisioned during bootstrap                     |

> | **Kubelet Cert**          | External system | CABPK         | node               | Handled via Kubernetes CSR API and custom signer |

Reviewer 3: I'm not entirely sure about those two lines, because what we are describing here is very specific to a provider/to how kubeadm works and this fact is not surfacing

Reviewer 3: Might be worth to add a generic version of this table, limited to what will be prescribed by the contract), and then a second version that show the specific implementation for KCP/CABPK

### External CA mode

When the flag is explicitly enabled, CA keys are no longer required in the CAPI management cluster and must not be stored or validated in secrets. By default, this feature remains disabled to avoid impacting users who do not rely on external CAs.
KCP/CABPK will not generate or manage CA private keys. The CA certificates needs to be provided by the user or an external system in Kubernetes secrets with the following naming convention:

- `<cluster-name>-etcd`
- `<cluster-name>-ca`
- `<cluster-name>-proxy`

Each Secret should follow this Kubernetes format:

```
apiVersion: v1
kind: Secret
metadata:
  name: <cluster-name>-<etcd|ca|proxy>
type: Opaque
data:
  ca.crt: <base64-encoded-CA-certificate>
```

> KCP/CABPK should wait for the CA certificates if they are not already provided. KCP/CABPK must not adopt these Secrets, as they are the responsibility of the user.

Reviewer 3: I'm a little bit concerned by 

> KCP/CABPK should wait for the CA certificates 

This implies that we are implementing also an orchestration process between KCP/CABPK and the external CA system, which is something we are not doing e.g in case of custom CA

TBH. I would prefer if this is remove from the proposal, because e.g. we already have other mechanism like the BeforeClusterCreate hooks that provide a similar orchestration workflow.

### Certificates necessary to KCP/CABPK

> #### Kubeadm and KCP requirements

Reviewer 3: I think that something else must change in CABPK, because it should stop sending keys to the machines as part of the cloud init secret.

The management of Certificate Authorities and internal component certificates falls under the scope of KCP and CABPK. KCP will not generate these keys in external CA mode and expects the user to provide the following secrets:

> - **Certificate Authorities:** The cluster CA, etcd CA, and front-proxy CA must be provided as secrets.

Reviewer 3: Let's also mention that the user should also take care of secret removal

> - **Etcd Client Certificate:** For clusters using stacked etcd, the user must provide the etcd client certificate in a secret named `<cluster-name>-apiserver-etcd-client`. For external etcd, KCP follows the existing logic for external etcd connectivity.

Reviewer 3: Does the user should take care of the lifecycle for this certificate (e.g. rotation, removal)? if yes, please call this out

#### Cluster API Contract and Kubeconfig

> The admin certificate used for the cluster kubeconfig is a core Cluster API requirement. This proposal adheres to the [Cluster API Control Plane Contract](https://cluster-api.sigs.k8s.io/developer/providers/contracts/control-plane#cluster-kubeconfig-management) by delegating the issuance of the `<cluster-name>-admin` secret to an external system while maintaining standard naming conventions. This ensures compatibility with the broader CAPI ecosystem.

Reviewer 3: I'm not entirely sure this proposal adheres to the current contract.

The current contract assigns to the control plane provider the resposnibility to create and maintain (rotate) the Kubeconfig secret, which is not true anymore in external CA mode, where this responsibility is assigned to an external system)

(TL;DR; I think we have to change the contract in https://cluster-api.sigs.k8s.io/developer/providers/contracts/control-plane#cluster-kubeconfig-management for this proposal to explicitly call this out).

> Certificates such as the etcd client certificate and the admin certificate for the kubeconfig are normally generated by KCP/CABPK. In external CA mode, these certificates are still needed, but their provisioning and lifecycle management become the responsibility of the user.

Reviewer 3: How other CAPI users are expected to do so?

> Kubeadm controllers will verify the presence and validity of external certificates using **two distinct conditions** on the `KubeadmControlPlane` resource:

Reviewer 3: I will try to keep it simple and use existing CertificatesAvailable in KCP/CABPK to surface issues

e.g. condition must be true when
- Certificate authorities are present and all have keys (custom CA or self signed)
- Certificates authorities are present and all don't have keys (custom CA or self signed) + user managed client cert are there as well

e.g. condition must be false when
- Failure in generating certificate authorities (self signed)
- Certificates authorities are present but in a mixed mode (some have keys, come other not)
- When in external CA mode, some of the user managed client cert are missing

Below the workflow that describe this proposal.

![workflow](images/external-cert-sign/workflow.png)

### User Stories

#### Enterprise PKI Integration

As a KCP/CABPK user in an enterprise environment, I want to delegate certificate signing to our corporate CA, so that CA keys never reside in the management cluster and certificate issuance complies with organizational security policies.

#### Flexible Bootstrap Provider Support

As a KCP/CABPK user I want that it provides all necessary configuration to bootstrap providers to be compatible with external signing, so that I can maintain secure certificate flows regardless of the underlying bootstrap mechanism.

### Security Model

The External CA / External Certificate Signing feature is designed to minimize exposure of CA private keys and improve the overall security posture of Cluster API clusters. The security model includes the following key points:

- CA private keys never leave the external CA system.
- Cluster API and management clusters do not store CA keys in secrets.

## Upgrade Strategy

> The `externalCA` feature can be enabled on an existing cluster that was originally created with the default internal CA, but it requires careful manual steps by the user. Cluster API does not automatically rotate or migrate CAs.

Reviewer 3: I would prefer to rule this out (non goal) to keep the scope of the proposal limited.
